Add advanced CPU topology settings in Basic Settings step

hstastna opened the pull request:
Fixes: https://github.com/oVirt/ovirt-web-ui/issues/1264

This PR adds support for editing advanced CPU topology (i.e. number of virtual sockets, cores per virtual socket, threads per core) in the _Basic Settings_ of the Create VM Wizard, by adding the "**Advanced CPU Topology Options**" that we can expand/collapse.
The chosen settings are also displayed in a popover next to the number of the virtual CPUs in the _Review_ report step of the Create VM Wizard.

Also the css style of the pf `ExpandCollapse` was slightly edited [to achieve the expected look](https://github.com/oVirt/ovirt-web-ui/pull/1279/commits/1575e874661782a4556d9bcbc1d6377f702083c0#diff-0cdb3888d313c777c2256655e5b4f867R236-R251) (same as before, extra space above the Adv. CPU Topology Options and above their content, proper align of the expand/collapse icon (^ V) with the rest of the component).

---

**After:**
_Basic settings_ step:
![adv1](https://user-images.githubusercontent.com/13417815/94066840-55654c80-fded-11ea-97c3-a110fb5b3a3d.png)
![adv2](https://user-images.githubusercontent.com/13417815/94066850-57c7a680-fded-11ea-9a6e-1f8930faee3a.png)
![adv3](https://user-images.githubusercontent.com/13417815/94066863-5a2a0080-fded-11ea-8a5a-59d046e81346.png)
_See also:_
https://github.com/oVirt/ovirt-web-ui/pull/1279#issuecomment-697999825
_Review_ step:
https://github.com/oVirt/ovirt-web-ui/pull/1279#issuecomment-688354076

bamsalem commented on src/components/CreateVmWizard/steps/SummaryReview.js:
```diff
@@ -59,7 +59,14 @@ const ReviewBasic = ({ id, dataCenters, clusters, isos, templates, operatingSyst
     <Item id={`${id}-cpus`} label={msg.cpus()}>
       { basic.cpus }
-      {/* TODO: Include topology (as a popover?) */}
+      <FieldLevelHelp
+        content={msg.totalCpuTooltip({
+          cores: basic.topology.cores,
+          sockets: basic.topology.sockets,
+          threads: basic.topology.threads,
+        })}
+        inline
+      />
```
It's not must, but I have issue #1094 which deal with tooltips so I'll suggest you to use `<InfoCircleIcon />` wrapped by `<OverlayTooltip>`

bamsalem reviewed (changes requested):
Hi Hilda it works great, but I have few suggestion:
there are several code duplicates, I suggests to refactor the code and extract it to functions/components and not duplicate code, that way future maintenance will be easier, WDYT?

hstastna commented on src/components/CreateVmWizard/steps/BasicSettings.js:
```diff
@@ -548,6 +602,60 @@ class BasicSettings extends React.Component {
+                  <FieldRow label={msg.virtualSockets()} id={`${idPrefix}-topology-sockets`} className={style['memory-input']}>
+                    <SelectBox
+                      id={`${idPrefix}-topology-sockets-edit`}
+                      items={vCpuTopologyDividers.sockets.map(i => ({
+                        id: i.toString(),
+                        value: i.toString(),
+                      }))}
+                      selected={data.topology.sockets.toString()}
+                      onChange={selectedId => { this.handleChange('topology', selectedId, { vcpu: 'sockets' }) }}
+                    />
+                  </FieldRow>
+                  <FieldRow label={msg.coresPerSockets()} id={`${idPrefix}-topology-cores`}>
+                    <SelectBox
+                      id={`${idPrefix}-topology-cores-edit`}
+                      items={vCpuTopologyDividers.cores.map(i => ({
+                        id: i.toString(),
+                        value: i.toString(),
+                      }))}
+                      selected={data.topology.cores.toString()}
+                      onChange={selectedId => { this.handleChange('topology', selectedId, { vcpu: 'cores' }) }}
+                    />
+                  </FieldRow>
+                  <FieldRow
+                    label={msg.threadsPerCores()}
+                    id={`${idPrefix}-topology-threads`}
+                    tooltip={
+                      isClusterPower8
+                        ? msg.recomendedPower8ValuesForThreads({ threads: maxNumOfThreads })
+                        : msg.recomendedValuesForThreads({ threads: maxNumOfThreads })
+                    }>
+                    <SelectBox
+                      id={`${idPrefix}-topology-threads-edit`}
+                      items={vCpuTopologyDividers.threads.map(i => ({
+                        id: i.toString(),
+                        value: i.toString(),
+                      }))}
+                      selected={data.topology.threads.toString()}
+                      onChange={selectedId => { this.handleChange('topology', selectedId, { vcpu: 'threads' }) }}
+                    />
+                  </FieldRow>
```
It's not almost the same code as in _DetailsCard_.

hstastna commented:
Update of displaying advanced CPU topology settings in the _Review_ step:
![summary_tooltip](https://user-images.githubusercontent.com/13417815/92396120-c4238400-f124-11ea-8856-e9a070f6649f.png)

The _OverlayTooltip_ and _InfoCircleIcon_ components were used instead of older _FieldLevelHelp_ with respect to the https://github.com/oVirt/ovirt-web-ui/issues/1094 (probably we'll need resolving some merge conflicts, after merging https://github.com/oVirt/ovirt-web-ui/pull/1289). The tooltip in the VM _Details_ card should be updated by https://github.com/oVirt/ovirt-web-ui/pull/1289, too.

rszwajko reviewed (dismissed):
1. In VM details the Advanced Options section covers also cloud-init. 
How about including both `Start virtual machine on creation` and `Enable Cloud-init/Sysprep` options in the wizard version?
2. Could we make the dropdowns the same width as `Total Virtual CPUs` input?

hstastna commented:
> 1. In VM details the Advanced Options section covers also cloud-init.
    How about including both Start virtual machine on creation and Enable Cloud-init/Sysprep options in the wizard version?

Could you please, explain this suggestion more? Thanks! :)

> Could we make the dropdowns the same width as Total Virtual CPUs input?

I spent some time solving this issue specifically but I haven't found any suitable solution, except ugly hacking the css or refactoring the whole Basic settings structure, or using different components that we already use. If we were okay without any expandable section (which complicates that), it would be already done. However, the UX folks strongly recommended to implement an expandable section (without any specific recommendations for drop downs width), similarly as I've implemented that. I can give it some more time but I'm not sure if it worth, regarding the time spent on it and regarding that this is a small cosmetic issue which has no impact on functionality.
Also note that those are drop downs, same as the other drop downs in the Basic settings. The Total Virtual CPUs input is a different component with arrow keys. I've implemented the drop downs for the number of sockets, cores, threads to keep the consistency of this feature with implementation in the VM Details card.

sjd78 reviewed (changes requested):
Adopting the topology editing looks very close.  A few gaps to close for default values and for layout.

hstastna commented on src/components/CreateVmWizard/steps/SummaryReview.js:
```diff
@@ -60,7 +62,17 @@ const ReviewBasic = ({ id, dataCenters, clusters, isos, templates, operatingSyst
+        })}
+        placement={'top'}
+      >
+        <InfoCircleIcon className={style['info-circle-icon']} />
```
I used the `info-circle-icon` name with respect to the class from one of the Ben's PRs (I don't remember which one exactly) where he created a style which is the same as we need here. The name was chosen probably because the class is used with `InfoCircleIcon` component. So I don't think that changing it to `cpu-toplogy-tooltip-icon` would be more accurate.

sjd78 replied:
Humm...sounds like something to address in the tooltips PR #1289 for consistency sake.  In that case, you're ok with leaving it as-is.

hstastna commented on src/components/CreateVmWizard/steps/BasicSettings.js:
```diff
@@ -547,6 +594,48 @@ class BasicSettings extends React.Component {
           </React.Fragment>
         }
+
+        {/* Advanced options */}
```
Well, I was thinking about improving displaying the _cloud-init/sysprep_ and _Start on creation_ checkboxes, too, in a separate PR. However, I haven't received clear info from the UX regarding this, just some assumptions, possibilities.

IMHO, it makes sense to place those checkboxes in the _Advanced Options_ expand section, too, because the actual position of those checkboxes isn't optimal - they make a feel that those options are closely connected and related to the _Optimized for_ drop down, and because I think those are advanced options, too. And also because I can see that _Cloud-init_ can be enabled specifically in the _Advanced Options_ of a VM Details card.

Yes, there's a possibility to rename the section to "Advanced CPU Topology" as you suggest. However, with renaming it, I think it will be less consistent with the same section containing CPU topology options in the VM _Details_ card, where I was inspired.

So I would choose the first option to solve this thing- to place those checkboxes into the _Advanced options_ section.

sjd78 replied:
It does make sense to visually group the 3 advanced things together (topology, start on create, cloud-init/sysprep).

However, I'd rather not put the start-on-create and cloud-init/sysprep options inside a click-to-see advanced section.  I'd rather see those two all the time, and then force a click to see the topology.

Let's limit scope creep.  The easiest thing to do right now is not worry about start-on-create and cloud-init/sysprep simply because they're not part of the issue this PR is solving.  We can raise a new issue and discuss if we actually want to do that later.  Something like "On the create VM wizard, consider grouping ____ in a default collapsed advanced options section".  We can then have a short meeting with people just to discuss if that is something we want to do or not.

hstastna replied:
Of course I didn't want to move `cloud-init/sysprep` and `Start on creation` options anywhere, within this PR :)

rszwajko replied:
>>  In VM details the Advanced Options section covers also cloud-init.
>>  How about including both Start virtual machine on creation and Enable Cloud-init/Sysprep options in the wizard 
>>  version?
> Could you please, explain this suggestion more? Thanks! :)

I do agree that we should stop feature creep :) But IMHO both checkboxes belong to advance section.

sjd78 replied:
> I do agree that we should stop feature creep :) But IMHO both checkboxes belong to advance section.

Great!  It'll be done with https://github.com/oVirt/ovirt-web-ui/issues/1297.  This PR's content is full.

hstastna commented on src/components/CreateVmWizard/steps/BasicSettings.js:
```diff
@@ -11,11 +11,14 @@ import {
 } from '_/components/utils'
 
 import { Form, FormControl, Checkbox } from 'patternfly-react'
+import ExpandCollapseSection from '_/components/ExpandCollapseSection'
```
Looks like we do have pf3 component for this:
https://www.patternfly.org/v3/pattern-library/forms-and-controls/expand-collapse-section/# (code not available 😞  )
https://patternfly-react-pf3.surge.sh/?path=/story/patternfly-react-forms-and-controls-expand-collapse--expandcollapse
https://github.com/patternfly/patternfly-react/tree/patternfly-3/packages/patternfly-react/src/components/ExpandCollapse

So I will make a change and swap it out in the VM _Details_ card as well, in a separate PR :) Thank you!

hstastna commented on src/components/CreateVmWizard/steps/BasicSettings.js:
```diff
@@ -547,6 +594,48 @@ class BasicSettings extends React.Component {
         }
+
+        {/* Advanced options */}
+        <ExpandCollapseSection id={`${idPrefix}-advanced-options`} sectionHeader={msg.advancedOptions()}>
```
Well, it seems that after I replaced the `ExpandCollapseSection` by pf3 component `ExpandCollapse`, wrapping the whole section in a Row and a Column and/or removing the `details-container` style didn't have any positive impact on how the whole section looks like. And I think I don't understand what you mean by "getting the collapse section to line up nice vertically".

Anyway, I think that the actual look of the whole section is fine, except that the drop downs are unnecessary wide (but on the other side, their width matches with the rest of drop downs in the Basic settings :D ).

sjd78 commented on src/components/CreateVmWizard/CreateVmWizard.js:
```diff
@@ -95,12 +95,18 @@ function getInitialState (clusters, templates, blankTemplateId, operatingSystems
+      topology: {
+        cores: parseInt(topology.get('cores'), 10),
+        sockets: parseInt(topology.get('sockets'), 10),
+        threads: parseInt(topology.get('threads'), 10),
```
I don't think the `parseInt()` is necessary on these fields.  If for some reason they are strings, it makes more sense to cast them to a number in the transform function:

https://github.com/oVirt/ovirt-web-ui/blob/2ad01a8885b1f07aa32a455eb1be10a9f62bce5a/src/ovirtapi/transform.js#L378-L385

sjd78 commented on src/components/CreateVmWizard/steps/BasicSettings.js:
```diff
@@ -547,6 +593,46 @@ class BasicSettings extends React.Component {
         }
+
+        {/* Advanced CPU Topology Options */}
+        <ExpandCollapse id={`${idPrefix}-advanced-options`} textCollapsed={msg.advancedCpuTopologyOptions()} textExpanded={msg.advancedCpuTopologyOptions()}>
```
Try it like:
```jsx
<Row>
  <Col offset={3} cols={7}>
    <ExpandCollapse ...>
      <Grid ...>
        <FieldRow>
          ...
        </FieldRow>
        ...
      </Grid>
    </ExpandCollapse>
  </Col>
</Row>
```

The outer-most `offset={3} cols={7}` will push the collapse section over to line up with the input fields.

The inner `FieldRow` components will still have 3 cols for the label and 7 cols for the input fields, just that the total space is now within the outer `Col`'s 7 columns.

Further, you could change `FieldRow` to make the component columns adjustable and set it to 4 or 5 if you want them to take less horizontal space.

hstastna replied:
https://github.com/oVirt/ovirt-web-ui/issues/1264#issuecomment-696258357 So I asked the UX about the expected look again and it seems that I will have to find a way how to shorten the CPU Topology drop downs but to leave the rest of the design as it is.

sjd78 replied:
Ok, 2 things.

1. The advanced section is below the regular field Grid, so it can be its own Grid.

2. You can use the `Col` to control width in a generic way and not have to tweak the css like you did.  That being said, straight CSS manipulation is more reasonable in other circumstances.

My suggestion (after adding `fieldCols` as a prop that defaults to 7 to the `FieldRow` component):
```jsx
    <Form ...>
      <Grid className={style['settings-container']}>
        ... the standard fields ...
      </Grid>

      <ExpandCollapse ...>
        <Grid className={style['settings-container']}>
          <FieldRow fieldCols={3} label={msg.virtualSockets()} id={`${idPrefix}-topology-sockets`}>
           <SelectBox
              ... 
            />
          </FieldRow>

          ...more advanced field rows...
        </Grid>
      </ExpandCollapse>
    </Form>
```

hstastna replied:
> The advanced section is below the regular field Grid, so it can be its own Grid.

I don't think it is below the regular field Grid, but I can move it below, however it seems that it doesn't help to achieve the expected look (other than straight CSS manipulation).

hstastna commented on src/components/CreateVmWizard/steps/BasicSettings.js:
```diff
@@ -258,6 +278,7 @@ class BasicSettings extends React.Component {
 
         changes.memory = template.get('memory') / (1024 ** 2) // stored in bytes, input in Mib
         changes.cpus = template.getIn([ 'cpu', 'vCPUs' ])
+        changes.topology = this.getTopologySettings(changes.cpus)
```
This is a very good point, Scott! So I've created a function `getTemplateTopology` for grabbing the topology from the template and getting the object containing only what we need. This function can be used also in one place in the_`CreateVmWizard.js_ :)

hstastna commented:
**Update:**
Regarding suggestions on making the width of the Advanced CPU Topology drop downs shorter, I've chosen a solution based on the editing the maximum width of the div elements with a specific ids. Even this solution depends on the ids, it's simple and works perfectly also when the drop down menu is opened, comparing to the other solutions with setting the number of cols of the components which would require not just a few css adjustments anyway to achieve the expected look.
The result:
![advv](https://user-images.githubusercontent.com/13417815/94074154-ff96a180-fdf8-11ea-9af3-3364b186fadb.png)

sjd78 commented on src/components/CreateVmWizard/steps/style.css:
```diff
@@ -227,3 +234,13 @@ table tbody tr td.kebab-menu-cell {
   padding-left: 10px;
 }
+
+.details-container {
```
Just use `.settings-container` - I think they're doing the same job.

sjd78 reviewed (dismissed):
The look and location is good-to-go.

A few comments relating to default topology and working with topology from templates.

I left the structure that I would use for the advanced section using `Col`'s `col` prop to control field width.  Using the Grid, be it ours or the one from PF, is far more common for controlling widths and alignment.

hstastna commented on src/components/CreateVmWizard/steps/BasicSettings.js:
```diff
@@ -243,6 +262,7 @@ class BasicSettings extends React.Component {
         changes.operatingSystemId = this.props.defaultValues.operatingSystemId
         changes.memory = this.props.defaultValues.memory
         changes.cpus = this.props.defaultValues.cpus
+        changes.topology = this.getTopologySettings(changes.cpus)
```
This does not really matter because `this.props.defaultValues.cpus` will always get the value `1` so both `this.getTopologySettings` and `this.props.defaultValues.topology` will come to the same result. But yeah, I can change that, just for consistency with the rest of the values here which are set to default, of course 😏

sjd78 replied:
Assuming that the default/blank template is unchanged, then yes. But...the default template can be changed!

hstastna replied:
👍🏽 Yes, it can (but I don't think it ever happens) :)

hstastna commented on src/components/CreateVmWizard/steps/BasicSettings.js:
```diff
@@ -287,6 +307,12 @@ class BasicSettings extends React.Component {
+        break
+
+      case 'topology': // number of sockets, cores or threads changed by the user in Advanced Options section
+        changes[field] = this.getTopologySettings(this.props.data.cpus, { [extra.vcpu]: parseInt(value, 10) })
```
What I can see above is removing the `parseInt'. If I do it, the feature stops working properly.

sjd78 replied:
Ah, it's the `+` so it looks like `+value` instead of `value`.

hstastna replied:
Yeah, I found that in the meantime 😄  👍🏽

hstastna commented:
> The advanced section is below the regular field Grid, so it can be its own Grid.

I don't think it is below the regular field Grid, but I can move it below, however it seems that it doesn't help to achieve the expected look (other than straight CSS manipulation).

hstastna commented:
**Update:**
I was trying to implement suggested [approach](https://github.com/oVirt/ovirt-web-ui/pull/1279#discussion_r493943125) by using `Col` component (with setting `offset` or number of `cols`) or `fieldCols` to the `FieldRow` component. Tried different values, combinations etc. but nothing worked as expected. I was able to achieve the expected result (proper aligning and width of the topology drop downs) only by adding some extra conditions to the `Col` component itself and extra css classes to other components. And at the end, using `Col` lost its point. I had to set the width using extra css anyway.

One of the reasons why this approach didn't not work well is the style of the `Col` component with the combination of the pf style of the `ExpandCollapse` component, especially the `.grid-column` class is problematic and unnecessary for our situation. But if we fixed that (by adding extra changes to our `Col` component or hacking css), we still just fixed proper aligning of the topology drop downs and their labels. Then there's still a problem with the width of the drop downs which is calculated slightly differently for the content of the `ExpandCollapse` component, so setting `cols` is useless. If we set the `cols`, we can only get closer to the expected width of the drop downs (if we decide not adding extra css styles and making it even more complicated). However the difference between their width and width of the _Memory_ and _Total vCPUs_ fields was relatively small, it got bigger, visible and annoying while changing size of the window. So I think the best is not using `Col` at all and to use the same approach as in the _src/components/VmDetails/cards/DetailsCard/index.js_ by putting the content of the `FieldRow` into extra divs with style we need (see my last commit). Or to use the very simple and working approach I suggested earlier by straight css manipulation: 
```
div[id^='create-vm-wizard-basic-topology-'] {
  max-width: 40%;
}
```

sjd78 reviewed (approved):
I updated the layout.

LGTM with that last change.

hstastna commented:
> I updated the layout.
> 
> LGTM with that last change.

Note that now the width of the drop downs is not 100% the same as Memory and Total vCPUs fields (it was with my previous solution) but if it is better now then okay.

sgratch commented:
@hstastna 
The advanced section looks great to me.
Please consider adding the "Threads per Core" tooltip, as exists for the VM's details card.

hstastna commented:
@sgratch Are you sure we need this tooltip even when the VM was not created yet?

michalskrivanek commented:
eh, if we have such a tooltip then at least fix the tooltip content please (separately)
That "Other recommended values..." - I would drop that sentence completely, it's misleading/confusing.

hstastna commented:
@sgratch The tooltip already added:
![tooltip](https://user-images.githubusercontent.com/13417815/95584619-511c6e80-0a3e-11eb-95fb-a29d3319645d.png)

The `FieldLevelHelp` component was used and "it should be refactored as part of #1289" as you've mentioned in your [comment](https://github.com/oVirt/ovirt-web-ui/pull/1279#issuecomment-705270974),

michalskrivanek commented:
no change since last approval other than fixing the tooltip content

isaranova commented:
LGTM
